lib/THC/THCTensorMath.cuh

```diff
@@ -23,4 +23,31 @@ __global__ void THCTensor_copyToDiagonal(T* a, T* b, ptrdiff_t start, ptrdiff_t
   }
 }
 
+#define CAT_ARRAY_KERNEL_BATCH_SIZE 2
```

Comment:
1) You can use the constant memory to store the pointers to all the arrays at the same time (assuming there are not tens of thousands of arrays you are cat'ing at the same time).

2) Pass in an array of `long`s which contain the values of the pointers to the arrays. The number of arrays can be equal to the number of blocks along y.